README
## Descrição

O DBIP é um projeto que visa identificar através do IP a origem da requisição.

## Requisitos

- Docker >= 24
- Docker compose >= 2.3

## Configuração

Descomente o arquivo `.env.example` para `.env` e adicione seus valores as variáveis conforme preferir.

## Executando os containers

Ao executar o comando abaixo, os containers: web, api, mongodb, mysql e redis, serão criados.

```bash
$ docker compose up
```

## DBIP
1 - Faça o download da base de dados [aqui](https://vturb-labs.s3.amazonaws.com/challenges/dbip.csv.gz).

2 - Faça o download do projeto [dbip-phpsrc-4.0.tgz](https://download.db-ip.com/free/dbip-phpsrc-4.0.tgz).

3 - Faça o download do [DBeaver](https://dbeaver.io/download/).

4 - Conecte-se ao `mysql` usando o `DBeaver` e crie a tabela conforme instrução sql abaixo:

```
CREATE TABLE `dbip_lookup` (
  `addr_type` enum('ipv4','ipv6') NOT NULL,
  `ip_start` varbinary(16) NOT NULL,
  `ip_end` varbinary(16) NOT NULL,
  `continent` char(2) NOT NULL,
  `country` char(2) NOT NULL,
  `stateprov_code` varchar(15) DEFAULT NULL,
  `stateprov` varchar(80) NOT NULL,
  `district` varchar(80) NOT NULL,
  `city` varchar(80) NOT NULL,
  `zipcode` varchar(20) NOT NULL,
  `latitude` float NOT NULL,
  `longitude` float NOT NULL,
  `geoname_id` int(10) unsigned DEFAULT NULL,
  `timezone_offset` float NOT NULL,
  `timezone_name` varchar(64) NOT NULL,
  `weather_code` varchar(10) NOT NULL,
  `isp_name` varchar(128) NOT NULL,
  `as_number` int(10) unsigned DEFAULT NULL,
  `usage_type` enum('corporate','consumer','hosting') DEFAULT NULL,
  `connection_type` enum('dialup','isdn','cable','dsl','fttx','wireless') DEFAULT NULL,
  `organization_name` varchar(128) NOT NULL,
  PRIMARY KEY (`addr_type`,`ip_start`)
) DEFAULT CHARSET=utf8mb4;
```

5 - Importando a base de dados DBIP

```bash
$ docker cp ./local/do/arquivo/dbip.csv.gz dbip-phpsrc-app-1:/tmp
$ docker exec -it dbip-phpsrc-app-1 ./import.php -f "/tmp/dbip.csv.gz" -d full
```


HISTORY.md
### 0.1.0 (2023-10-25)
- Versão inicial do projeto

### 0.1.1
- Adicionando cabeçalho para permitir CORS.

public/index.php
<?php

header("Content-Type: application/json; charset=utf-8");
header("Access-Control-Allow-Origin: *");

/**
 *
 * DB-IP.com database sample query code
 *
 * Copyright (C) 2018 db-ip.com
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

// Include the DB-IP class
require "../dbip.class.php";

$ip = htmlspecialchars($_GET['ip']);

if (!$ip) {
	http_response_code(400);
	echo json_encode(array(
		'message' => 'Parâmetro IP é obrigatório.'
	), JSON_UNESCAPED_UNICODE);
	exit();
}

try {
	// Connect to the database
	$db = new PDO("mysql:host=db;dbname=dbip", getenv('MYSQL_USER'), getenv('MYSQL_PASSWORD'));

	// Alternatively connect to MySQL using the old interface
	// Comment the PDO statement above and uncomment the mysql_ calls
	// below if your PHP installation doesn't support PDO :
	// $db = mysql_connect("localhost", "root", "");
	// mysql_select_db("test", $db);

	// Instanciate a new DBIP object with the database connection
	$dbip = new DBIP($db);

	// Alternatively instanciate a DBIP_MySQL object
	// Comment the new statement above and uncomment below if your PHP
	// installation doesn't support PDO :
	// $dbip = new DBIPMySQL($db);

	// Lookup an IP address
	$inf = $dbip->lookup($ip);

	// Show the associated country
	echo json_encode($inf, JSON_UNESCAPED_UNICODE);
} catch (DBIPException $e) {
	echo "error: {$e->getMessage()}\n";
	exit(1);
}

src/import.php
#!/usr/bin/env php
<?php

/**
 *
 * DB-IP.com database import script
 *
 * Copyright (C) 2018 db-ip.com
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

require "dbip.class.php";

$opts = getopt("f:d:t:b:u:p:");

$filename = @$opts["f"];
$type = @$opts["d"];

$dbname = @$opts["b"]
or $dbname = "dbip";

$table = @$opts["t"]
or $table = "dbip_lookup";

$username = getenv('MYSQL_USER');

$password = getenv('MYSQL_PASSWORD');

if (!isset($type) && preg_match('/dbip-(country-lite|city-lite|country|location|isp|full)/i', $filename, $res)) {
	$type = $res[1];
}

if (!isset($filename) || !isset($type)) {
	die("usage: {$argv[0]} -f <filename.csv[.gz]> [-d <country-lite|city-lite|country|location|isp|full>] [-b <database_name>] [-t <table_name>] [-u <username>] [-p <password>]\n");
}

switch (strtolower($type)) {
	case "country-lite":	$dbtype = DBIP::TYPE_COUNTRY_LITE;	break;
	case "city-lite":		$dbtype = DBIP::TYPE_CITY_LITE;		break;
	case "country":			$dbtype = DBIP::TYPE_COUNTRY;		break;
	case "location":		$dbtype = DBIP::TYPE_LOCATION;		break;
	case "isp":				$dbtype = DBIP::TYPE_ISP;			break;
	case "full":			$dbtype = DBIP::TYPE_FULL;			break;
	default:				echo "invalid database type\n"; exit(1);
}

try {
		// Connect to the database
        $db = new PDO("mysql:host=db;dbname={$dbname};charset=utf8mb4", $username, $password);
        // Alternatively connect to MySQL using the old interface
        // Comment the PDO statement above and uncomment the mysql_ calls
        // below if your PHP installation doesn't support PDO :
        // $db = mysql_connect("localhost", $username, $password);
        // mysql_select_db($dbname, $db);

        // Instanciate a new DBIP object with the database connection
        $dbip = new DBIP($db);
        // Alternatively instanciate a DBIP_MySQL object
        // Comment the new statement above and uncomment below if your PHP
        // installation doesn't support PDO :
        // $dbip = new DBIPMySQL($db);

	$nrecs = $dbip->importFromCsv($filename, $dbtype, $table, function($progress) {
		echo "\r{$progress} ...";
	});
	echo "\rfinished importing " . number_format($nrecs) . " records\n";
} catch (DBIPException $e) {
	echo "error: {$e->getMessage()}\n";
}